packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol

```diff
@@ -0,0 +1,389 @@
+    /// as this prevents silly footgunning.
+    /// See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+    uint256 internal _reserve5;
+    uint256 private _reserve6;
+    uint256 private _reserve7;
+    uint256 private _reserve8;
+    uint256 private _reserve9;
+    uint256 private _reserve10;
+    uint256 private _reserve11;
+    uint256 private _reserve12;
+    uint256 private _reserve13;
+    uint256 private _reserve14;
+    uint256 private _reserve15;
+    uint256 private _reserve16;
+    uint256 private _reserve17;
+    uint256 private _reserve18;
+    uint256 private _reserve19;
+    uint256 private _reserve20;
+    uint256 internal _reserve21;
+
+    constructor(
+        IConstantFlowAgreementV1 cfaV1Contract
+    ) {
+        CONSTANT_FLOW_AGREEMENT_V1 = cfaV1Contract;
+    }
+
+    function initialize(
+        ISuperToken superTokenContract,
+        string memory nftName,
+        string memory nftSymbol
+    )
+        external
+        initializer // OpenZeppelin Initializable
+    {
+        superToken = superTokenContract;
+        _name = nftName;
+        _symbol = nftSymbol;
+    }
+
+    function updateCode(address newAddress) external override {
+        if (msg.sender != address(superToken.getHost())) {
+            revert CFA_NFT_ONLY_HOST();
+        }
+
+        UUPSProxiable._updateCodeAddress(newAddress);
+    }
+
+    /// @notice Emits the MetadataUpdate event with `tokenId` as the argument.
+    /// @dev Callable by anyone.
+    /// @param tokenId the token id to trigger a metaupdate for
+    function triggerMetadataUpdate(uint256 tokenId) external {
+        _triggerMetadataUpdate(tokenId);
+    }
+
+    /// @notice This contract supports IERC165, IERC721 and IERC721Metadata
+    /// @dev This is part of the Standard Interface Detection EIP: https://eips.ethereum.org/EIPS/eip-165
+    /// @param interfaceId the XOR of all function selectors in the interface
+    /// @return boolean true if the interface is supported
+    /// @inheritdoc IERC165
+    function supportsInterface(
+        bytes4 interfaceId
+    ) external pure virtual override returns (bool) {
+        return
+            interfaceId == 0x01ffc9a7 || // ERC165 Interface ID for ERC165
+            interfaceId == 0x80ac58cd || // ERC165 Interface ID for ERC721
+            interfaceId == 0x5b5e139f; // ERC165 Interface ID for ERC721Metadata
+    }
+
+    /// @inheritdoc IERC721
+    function ownerOf(
+        uint256 tokenId
+    ) public view virtual override returns (address) {
+        address owner = _ownerOf(tokenId);
+        if (owner == address(0)) {
+            revert CFA_NFT_INVALID_TOKEN_ID();
+        }
+        return owner;
+    }
+
+    /// @notice Returns a hardcoded balance of 1
+    /// @dev We always return 1 to avoid the need for additional mapping
+    /// @return balance = 1
+    function balanceOf(
+        address // owner
+    ) external pure returns (uint256 balance) {
+        balance = 1;
```

Comment:
Could theoretically use a stored counter solution here to get an actual count. But heard we didn't do it because of gas reasons.

Reply:
Also, since we can't force retro mint, the counter wouldn't reflect correct numbers for existing ones.

Reply:
I think it would reflect the correct number for this particular NFT.

What happens to the old NFTs btw?